### PR TITLE
Issue 176 implement notification trigger mechanism: Investor Follows Startup Projects

### PR DESCRIPTION
**Pull request by heeylena**
In this task, I implemented a comprehensive notification trigger mechanism for when investors follow startup projects.  
Here's a summary of what has been completed:

### 1. FollowedProject Model (`investors/models.py`)
- Created a new model to track investor–project relationships.

### 2. Django Signals (`communications/signals.py`)
- Added `_connect_followed_project_signal()` function.
- Triggers notifications when `FollowedProject` is created.
- Includes deduplication logic to prevent duplicate notifications.
- Uses `transaction.on_commit()` for reliability.
- Handles edge cases (missing users, self-follows, etc.).

### 3. API Endpoints (`investors/views.py`)
- **FollowProjectView**: Simple endpoint to follow a project  
  `POST /api/v1/investors/projects/{id}/follow/`
- **UnfollowProjectView**: Endpoint to unfollow a project  
  `DELETE /api/v1/investors/projects/{id}/unfollow/`
- Proper authentication and permission checks.

### 4. Serializers (`investors/serializers/investor.py`)
- **FollowedProjectSerializer**: Handles validation and serialization.
- Includes nested project information (title, description, startup name).
- Prevents self-following through validation.

### 5. Database Migration (`investors/migrations/0007_followedproject.py`)
- Creates the `followed_projects` table.
- Adds indexes and constraints.
- Updates the Investor model with the many-to-many relationship.

### 6. Comprehensive Tests
- **Notification Tests** (`tests/communications/test_project_follow_notifications.py`):  
  Tests signal triggers and notification creation.
- **API Tests** (`tests/investors/test_followed_project_api.py`):  
  Tests all API endpoints and edge cases.

### 7. URL Routing (`investors/urls.py`)
- Added routes for the `FollowedProjectViewSet`.
- Added routes for follow/unfollow endpoints.

### 8. Admin Interface (`investors/admin.py`)
- Added Django admin interface for managing `FollowedProject` instances.
- Includes search, filtering, and proper field organization.

### CODE REVIEW

File: investors/models.py [hunk 1]
```diff
@@ -151,3 +160,67 @@ class Meta:
...
+        if self.notes is None:
+            self.notes = ""
+
+        # Prevent investors from following their own projects
+        inv_user_id = self.investor.user_id if getattr(self, 'investor_id', None) else None
+        proj_startup_user_id = None
+        if getattr(self, 'project_id', None):
+            project = getattr(self, 'project', None)
+            if project and hasattr(project, 'startup') and hasattr(project.startup, 'user_id'):
+                proj_startup_user_id = project.startup.user_id
```
Comment (mehalyna) [inline investors/models.py:206]: The method accesses fields that may not be loaded, risking extra queries or AttributeError. Refactor to ensure related objects are loaded or check existence before access.
```
if self.project_id and hasattr(self, 'project') and getattr(self.project, 'startup', None):
    proj_startup_user_id = getattr(self.project.startup, 'user_id', None)
```

File: investors/views.py [hunk 2]
```diff
@@ -215,6 +218,156 @@ def perform_destroy(self, instance):
...
+        ).exists():
+            logger.warning(
+                "FollowedProject create failed: duplicate",
+                extra={"investor_id": user.investor.pk, "project_id": project_id, "by_user": user.pk},
+            )
+            existing = FollowedProject.objects.get(investor=user.investor, project_id=project_id)
+            return Response(
+                FollowedProjectSerializer(existing).data,
+                status=status.HTTP_200_OK
+            )
```
Comment (mehalyna) [inline investors/views.py:264]: The view checks for duplicates with a DB query manually before serializer validation, which is redundant with unique constraints. Rely on serializer validation and handle IntegrityError for clean logic.
```
try:
    serializer.is_valid(raise_exception=True)
    self.perform_create(serializer)
except IntegrityError:
    # handle duplicate gracefully
```

File: investors/serializers/investor.py [hunk 3]
```diff
@@ -180,3 +188,51 @@ class Meta:
...
+        
+        investor = request.user.investor
+        
+        # Check if project exists and is accessible
+        if not value:
+            raise serializers.ValidationError("Project is required.")
+        
+        # Prevent following own projects
+        if hasattr(value, 'startup') and value.startup.user_id == request.user.pk:
+            raise serializers.ValidationError("You cannot follow your own project.")
```
Comment (mehalyna) [inline investors/serializers/investor.py:225]: The validator compares value.startup.user_id with request.user.pk directly, which can cause issues if startup or user is None. Add explicit None checks before comparison.
```
if hasattr(value, 'startup') and value.startup and value.startup.user_id == request.user.pk:
    raise serializers.ValidationError("You cannot follow your own project.")
```

File: investors/views.py [hunk 4]
```diff
@@ -215,6 +218,156 @@ def perform_destroy(self, instance):
...
+            raise ValidationError({"project": "You cannot follow your own project."})
+
+        try:
+            instance = serializer.save(investor=user.investor)
+        except IntegrityError:
+            logger.warning(
+                "FollowedProject create failed: duplicate",
+                extra={"investor_id": user.investor.pk, "project_id": project.pk, "by_user": user.pk},
+            )
+            raise ValidationError({"non_field_errors": ["Already followed."]})
```
Comment (mehalyna) [inline investors/views.py:320]: The exception handler for IntegrityError does not specify the exception to catch, which may mask other errors. Import and catch IntegrityError explicitly.
```
from django.db import IntegrityError

try:
    instance = serializer.save(investor=user.investor)
except IntegrityError:
    # handle duplicate
```
